packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -0,0 +1,160 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type { TreeNode } from "../simple-tree/index.js";
+
+/**
+ * A special object that signifies when a SharedTree {@link RunTransaction | transaction} should "roll back".
+ * @public
+ */
+export const rollback = Symbol("SharedTree Transaction Rollback");
+
+/**
+ * A requirement for a SharedTree transaction to succeed.
+ * @remarks Transaction constraints are useful for validating that the state of the tree meets some requirement when a transaction runs.
+ * In general, when running a transaction a client can validate their tree state in whatever way they wish and decide to either proceed with the transaction or not.
+ * However, they cannot know what the tree state will be when the transaction is _sequenced_.
+ * There may have been any number of edits from other clients that get sequenced before the transaction is eventually sequenced.
+ * Constraints provide a way to validate the tree state after the transaction has been sequenced and abort the transaction if the constraints are not met.
+ * All clients will validate the constraints of a transaction when it is sequenced, so all clients will agree on whether the transaction succeeds or not.
+ * @public
+ */
+export type TransactionConstraint = NodeInDocumentConstraint; // TODO: Add more constraint types here
+
+/**
+ * A transaction {@link TransactionConstraint | constraint} which requires that the given node exists in the tree.
+ * @remarks The node must be in the document (its {@link TreeStatus | status} must be {@link TreeStatus.InDocument | InDocument}) to qualify as "existing".
+ * @public
+ */
+export interface NodeInDocumentConstraint {
+	readonly type: "nodeInDocument";
+	readonly node: TreeNode;
+}
+
+/**
+ * The successful outcome of a transaction, i.e. the transaction should continue.
```

Comment:
If we keep this "continue" name, we should clarify here that when this is returned from the outer-most transaction scope, the transaction will end.

Reply:
Not sure I follow. I used the word "continue" to convey from the user's perspective that the user part or body of the transaction was successful and the system part of the transaction process can continue. But I guess that's a little confusing when it comes to nested transaction.
Maybe we can use some other word. I thought of "run" but that somehow doesn't convey the right idea. Also, it has to be apposite of "abort". Maybe "confirm" or "complete"?

Reply:
Ah, I can see how "continue" makes sense in that light. What I find the most confusing about it is that I would expect it to mean "please keep the transaction open for more changes that will add later (either in this JS turn or a later one)" which is something we used to support and might support again.

It's surprisingly hard to find something good here.
* "commit" seems odd when it comes to nested transactions. It also is confusing when you think about the fact that changes are only truly committed once they've been sequenced (but perhaps we can say the changes are committed but not pushed)
* "apply" seems wrong because we apply changes as we go so all changes are already applied.
* "ok" is not bad.

I think my preference might be to just not have a `result` field on this type or at least make sure that the user doesn't have to populate it (we may still want to field declared for the sake of using the type as a discriminated union).

Otherwise, "commit" seems okay. Either way I'm leaning for the type to be named `CommitTransaction`.